architecture/2025-03-upgrade-feature-spec.md

```diff
@@ -0,0 +1,221 @@
+- Ensure user data safety during the upgrade process by taking snapshots and providing rollback capabilities.
+- Provide clear documentation and guidance for users performing upgrades.
+- Maintain system performance during the upgrade process to avoid significant impact on running applications.
+
+### Non-goals (out of scope)
+
+- Downgrades are not supported. The upgrade process is designed to move forward to newer versions only.
+- The upgrade process does not include features for managing upgrades across multiple clusters simultaneously as Radius doesn't support multiple clusters per installation as of March 2025.
+- Upgrading major versions of dependencies like Postgres, Dapr, or Contour is not covered by this process. These upgrades should be handled separately following their respective guidelines.
+
+## User profile and challenges
+
+The primary users of this feature are system administrators and DevOps engineers responsible for maintaining and upgrading the Radius platform.
+
+### User persona(s)
+
+- **System Administrator**: Responsible for managing the infrastructure and ensuring the smooth operation of the Radius platform.
+- **DevOps Engineer**: Focused on automating deployment processes and maintaining the CI/CD pipeline.
+
+### Challenge(s) faced by the user
+
+- As of March 2025, Radius does not support in-place upgrades or upgrades via the Radius CLI. Users can achieve this using Helm, but this method does not update the local Radius CLI version to the desired version. The current steps are:
+  - `helm install radius oci://ghcr.io/radius-project/helm-chart/radius --version 0.43.0`
+  - `helm upgrade radius oci://ghcr.io/radius-project/helm-chart/radius --version 0.44.0`
+- This approach is cumbersome and does not provide a good user experience. Additionally, it does not automatically install dependencies like Contour and Dapr, requiring users to handle these installations manually.
+- Another way to upgrade Radius to the desired version involves the following steps:
+  - Uninstall the existing Radius installation by running `rad uninstall kubernetes`. This command doesn't delete any user data or the `radius-system` namespace.
+  - Download and install the desired version locally. Ensure it is the correct version by running `rad version`.
+  - Run `rad install kubernetes` or `rad init` to finalize the upgrade.
+- This process is also not as seamless as a `rad upgrade kubernetes --version desired.version` command could be.
+
+### Positive user outcome
+
+- By running the `rad upgrade kubernetes --version desired.version` command, users will be able to initiate the upgrade process seamlessly.
+  - The desired version must be higher than the current version.
+- By setting a global flag indicating that an upgrade is in progress, we ensure that users do not start data-changing commands like `rad deploy app.bicep` or `rad delete app` in a different terminal.
```

Comment:
Is the plan to block any mutating operations on the server side? Where is this flag stored?

Reply:
Yes, the plan is to let other users (terminals, etc.) know that there is an upgrade going on so that they shouldn't kick off any operations that would change the data.

Not sure where the flag will be stored as of now but shouldn't be a big problem. (I hope)